ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/Node.hs

```diff
@@ -180,6 +181,9 @@ data RunNodeArgs m addrNTN addrNTC blk (p2p :: Diffusion.P2P) = RunNodeArgs {
 
       -- | Network PeerSharing miniprotocol willingness flag
     , rnPeerSharing :: PeerSharing
+
+      -- | An event handler to trigger custom action when ledger events are emitted.
```

Comment:
This current diff is invoking this whenever chain selection validates a block. It's unclear to me that that is the most useful thing to do.

As such, would you reword and expand this comment to specify (in plain terms that your anticipated users of this feature would be likely to correctly interpret---not just eg Consensus team members) exactly which ledger events will be passed to this handler?

Reply:
For example, I would want to double-check whether the events for blocks on A be emitted twice times if a node switches from A to B and then back to an extension of A. And is that the desired behavior?

Does the client need to be informed of blocks whose events were previously emitted being rolled back? Or are they suppose to track that on their own, based on the emitted hashes and slots?

Reply:
Do they definitely need events for the volatile blocks? Or could you instead side-step these questions by only even emitting events for a block when it becomes part of the immutable chain?

Reply:
> @nfrisby This current diff is invoking this whenever chain selection validates a block.

Correct, although the validation is what's driving the addition of blocks to the volatile db. The indented behavior here really is to emit an event when a block is added to our local db. 

> @nfrisby exactly which ledger events will be passed to this handler?

Not sure to understand the question. This will pass ALL ledger events as emitted by the ledger. What events depend on the era and the relative time within the epoch. Are you asking to list all those in the comments here? (at the risk of creating discrepancy as soon as the ledger adds / removes events?) -- We could perhaps link to the relevant section of the ledger?

> @nfrisby For example, I would want to double-check whether the events for blocks on A be emitted twice times if a node switches from A to B and then back to an extension of A. And is that the desired behavior?

That is the desired behavior. In this model, we let clients cope with this and that's why events are emitted alongside a block header hash and a slot number. With these added information, clients should be able to figure out by themselves that a chain switch has occurred and that they need to rollback some of their internal states. 

> @nfrisby  Do they definitely need events for the volatile blocks? Or could you instead side-step these questions by only even emitting events for a block when it becomes part of the immutable chain?

That's a good question which can only be answered by use-cases I believe. Only relying on the immutable means that we are always lagging ~18h in the past. This may or may not sufficient depending on the use case (and I need to give it some thoughts for my own use case 🤔 ...). I believe that emitting events for volatile blocks is the most flexible. If clients need the immutability, they can wait for k blocks. The opposite isn't possible if we only emit events for immutable blocks.

Reply:
> > exactly which ledger events will be passed to this handler?

> Not sure to understand the question.

I meant the ledger events from "exactly which" blocks will be passed to the handler. EG you could say "every block the Consensus Layer chooses to validate". However, that's probably too opaque for the end-user. So how to describe it more clearly than that?

Reply:
I think dispatching events to the consumer at the point where chain selection occurs is the behaviour we want. In this current design, we want something that's minimally intrusive and therefore lay the burden of storing, indexing, and managing events to the client, providing them "all" the relevant information, eg. the block/slot and the event itself.
In any case, clients of a blockchain need to be able to understand and correctly handle rollbacks and forks. The block/slot index makes it easy to correlate the data of a block as provided by `ChainSync` and the events this block generated.
Not sure if this answers your questions @nfrisby  :)

Reply:
Yeah, I think so. I can imagine the intended spec is something along the lines:

> An event handler that will be called with the ledger events arising from the validation of a block that was ever one of the k latest blocks on the node's selection. Note that this means the stream of ledger events will be affected by roll backs, and that it will exclude events from blocks that were older than k (eg the entire historical chain) when the node started.

At the moment, though it will currently also include events from blocks that were never (yet) selected, since currently the node does validate blocks from the near-future even before they are selectable 😬 (that's something we're working to stop).

Reply:
@nfrisby we've been quite careful to not wire in the event handler in functions like [`chainSelectionForFutureBlocks`](https://github.com/CardanoSolutions/ouroboros-consensus/commit/852ae3de3575244d2065c4d8ab2633e5f0bcd846#diff-61e096d18a177444b21c891d73e650ebdb92a7b4524886447e221dc9bb99dfa0R395-R406) and [`addBlockAsync`](https://github.com/CardanoSolutions/ouroboros-consensus/commit/852ae3de3575244d2065c4d8ab2633e5f0bcd846#diff-61e096d18a177444b21c891d73e650ebdb92a7b4524886447e221dc9bb99dfa0R239-R246) to avoid precisely what you're describing. 

Is this "pointless" in the sense that the selection of future blocks will still happen through other execution paths?